corehq/ex-submodules/pillowtop/pillow/interface.py

```diff
@@ -184,6 +184,10 @@ def process_offset_chunk(chunk, context):
                 else:
                     self._update_checkpoint(None, None)
             process_offset_chunk(changes_chunk, context)
+            if forever:
+                if change:
+                    self._update_checkpoint(change, context)
+                self.process_changes(since=self.get_last_checkpoint_sequence(), forever=forever)
```

Comment:
shouldn't this be outside of the try except block?

Reply:
I don't think so. The modifications to `iter_changes` means the `for` loop will just end, which is when we want this to fire. So it needs to be outside the `for` loop, but no exception should be raised

Reply:
> but no exception should be raised

I was suggesting putting it after the try except since it's a recursive call to the same method. It seems cleaner to me to have it outside and I would think it's a bit 'cheaper' since the recursion wouldn't be inside the try except context. Either way it's not a blocker.

```python
checkpoint_reset = False
try:
    for change ....:
        ....
except PillowtopCheckpointReset:
    process_offset_chunk(changes_chunk, context)
    checkpoint_reset = True

if forever or checkpoint_reset:
    if change and not checkpoint_reset:
        self._update_checkpoint(change, context)
    self.process_changes(since=self.get_last_checkpoint_sequence(), forever=forever)
```

Reply:
Oh yeah, I didn't understand what you were asking before, but that makes a ton of sense. Will make that update